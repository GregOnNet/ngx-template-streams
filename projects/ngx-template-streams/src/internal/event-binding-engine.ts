import { some } from 'fp-ts/lib/Option';
import { escapeQuotes, escapeEventPayload } from '../utils/template-helpers';
import { INTERNAL_PREFIX } from './constants';
import { parseEventBindings } from './event-binding-parser';

type TemplateOperation = (template: string) => string;

export function updateEventBindings(template: string, isInlineTemplate = false) {
  return parseEventBindings(template)
    .chain(transformEventBindings(template, isInlineTemplate))
    .getOrElse(template);
}

function transformEventBindings(template: string, isInlineTemplate: boolean) {
  return (eventBindings: Array<RegExpExecArray>) => {
    return getTemplateOperations(eventBindings, isInlineTemplate).map(executeOperations(template));
  };
}

function getTemplateOperations(eventBindings: Array<RegExpExecArray>, isInlineTemplate: boolean) {
  const operations = eventBindings.map(([fullMatch, event, streamName, eventPayload]) => {
    const eventBinding = createEventBinding(event, streamName, eventPayload, isInlineTemplate);
    return createEventBindingOperation(fullMatch, eventBinding);
  });

  return some(operations);
}

function executeOperations(template: string) {
  return (operations: Array<TemplateOperation>) => {
    return operations.reduce<string>((acc: string, operation: TemplateOperation) => {
      return operation(acc);
    }, template);
  };
}

function createEventBinding(event: string, streamName: string, eventPayload = '$event', isInlineTemplate: boolean) {
  const eventBinding = `(${event})="${INTERNAL_PREFIX}${streamName}.next(${escapeEventPayload(eventPayload)})"`;
  return isInlineTemplate ? eventBinding : escapeQuotes(eventBinding);
}

function createEventBindingOperation(fullMatch: string, replacement: string): TemplateOperation {
  return (s: string) => {
    return s.replace(fullMatch, replacement);
  };
}
