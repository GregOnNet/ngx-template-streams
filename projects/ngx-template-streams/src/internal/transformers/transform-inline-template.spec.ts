import { resolve } from 'path';
import { fixtures, getFixtureFactory } from '../../testing/test-helpers';
import { transformFile } from '../../testing/ts-compiler';
import * as engine from '../event-binding-engine';
import * as transformer from './transform-inline-template';

const getComponentFixture = getFixtureFactory('component');
const getFixture = getFixtureFactory();

describe('inlineTemplateTransformer', () => {
  let transformerSpy: jest.SpyInstance;
  let updateEventBindingsSpy: jest.SpyInstance;

  beforeEach(() => {
    transformerSpy = jest.spyOn(transformer, 'inlineTemplateTransformer');
    updateEventBindingsSpy = jest.spyOn(engine, 'updateEventBindings');
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should apply transformation to inline template', () => {
    const { sourceFile } = getComponentFixture('inline-template');

    const output = transformFile(resolve(fixtures, sourceFile), [transformer.inlineTemplateTransformer]);

    expect(output).toMatchSnapshot();
    expect(transformerSpy).toHaveBeenCalled();
    expect(updateEventBindingsSpy).toHaveBeenCalled();
  });

  it('should not apply transformation if component has an external template', () => {
    const { sourceFile } = getComponentFixture('external-template');

    const output = transformFile(resolve(fixtures, sourceFile), [transformer.inlineTemplateTransformer]);

    expect(output).toMatchSnapshot();
    expect(updateEventBindingsSpy).not.toHaveBeenCalled();
  });

  it('should not apply transformation to empty file', () => {
    const { sourceFile } = getFixture('empty-file');

    const output = transformFile(resolve(fixtures, sourceFile), [transformer.inlineTemplateTransformer]);

    expect(output).toMatchSnapshot();
    expect(updateEventBindingsSpy).not.toHaveBeenCalled();
  });

  it('should transform inline template of multiple classes in a single file', () => {
    const { sourceFile } = getFixture('multiple-classes');

    const output = transformFile(resolve(fixtures, sourceFile), [transformer.inlineTemplateTransformer]);

    expect(output).toMatchSnapshot();
    expect(updateEventBindingsSpy).toHaveBeenCalledTimes(2);
  });
});
