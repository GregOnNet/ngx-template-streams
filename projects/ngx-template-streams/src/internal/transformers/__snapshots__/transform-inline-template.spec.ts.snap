// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inlineTemplateTransformer should apply transformation to inline template 1`] = `
"import { Component } from '@angular/core';
@Component({
    selector: 'test',
    template: \`<button (click)=\\"__click$.next('the payload')\\">Click Me</button>\`,
    styleUrls: ['./app.component.css']
})
export class TestComponent {
}
"
`;

exports[`inlineTemplateTransformer should not apply transformation if component has an external template 1`] = `
"import { Component } from '@angular/core';
@Component({
    selector: 'test',
    templateUrl: './some/path',
    styleUrls: ['./app.component.css']
})
export class TestComponent {
}
"
`;

exports[`inlineTemplateTransformer should not apply transformation to empty file 1`] = `""`;

exports[`inlineTemplateTransformer should transform inline template of multiple classes in a single file 1`] = `
"import { Component } from '@angular/core';
import { ObservableEvent } from 'ngx-template-streams';
import { Observable } from 'rxjs';
@Component({
    selector: 'a',
    template: \`<button (click)=\\"__a$.next($event)\\">Test<button>\`
})
export class ClassA {
    @ObservableEvent()
    a$: Observable<any>;
}
@Component({
    selector: 'b',
    template: \`<button (click)=\\"__b$.next($event)\\">Test<button>\`
})
export class ClassB {
    @ObservableEvent()
    b$: Observable<any>;
}
"
`;
